File: sample-videochat-webrtc/sample-videochat-webrtc/Components/Call/CallHelper/CallInfo.m
//
//  CallInfo.m
//  sample-videochat-webrtc
//
//  Created by Injoit on 21.09.2020.
//  Copyright © 2020 QuickBlox Team. All rights reserved.
//

#import "CallInfo.h"
#import "Profile.h"
#import <QuickbloxWebRTC/QuickbloxWebRTC.h>

@interface CallInfo() <QBRTCClientDelegate>

/// Sorted participants ids.
@property (nonatomic, strong) NSArray<NSNumber *>*list;
/// The call participants details where the key is a user id.
@property (nonatomic, strong) NSDictionary<NSNumber *, CallParticipant *> *cache;
@property (nonatomic, strong) Profile *profile;
@property (nonatomic, assign) CallDirection direction;
@property (nonatomic, strong) NSString *callId;

@end

@implementation CallInfo
//MARK: - Class Methods
+ (CallInfo *)callInfoWithCallID:(NSString *)callId
                         members:(NSDictionary<NSNumber *, NSString *>*)members
                       direction:(CallDirection)direction {
    return [[CallInfo alloc] initWithCallID:callId members:members direction:direction];
}

//MARK - Setup
- (NSNumber *)localParticipantId {
    return @(self.profile.ID);
}

- (NSArray<CallParticipant *> *)interlocutors {
    NSMutableArray<CallParticipant *> *interlocutors = [NSMutableArray array];
    for (NSNumber *userID in self.list) {
        if (userID.unsignedIntValue == self.profile.ID) { continue; }
        CallParticipant *participant = [self participantWithId:userID];
        if (participant) {
            [interlocutors addObject:participant];
        }
    }
    return  interlocutors;
}

- (NSArray<CallParticipant *> *)participants {
    NSMutableArray<CallParticipant *> *participants = [NSMutableArray array];
    for (NSNumber *userID in self.list) {
        CallParticipant *participant = [self participantWithId:userID];
        if (participant) {
            [participants addObject:participant];
        }
    }
    return  participants;
}

//MARK: - Life Cycle
- (void)clear {
    self.onChangedState = nil;
    self.onChangedBitrate = nil;
    self.onUpdatedParticipant = nil;
    [QBRTCClient.instance removeDelegate:self];
}

- (instancetype)initWithCallID:(NSString *)callId
                        members:(NSDictionary<NSNumber *, NSString *>*)members
                      direction:(CallDirection)direction {
    self = [super init];
    if (self) {
        
        [QBRTCClient.instance addDelegate:self];
        _callId = callId;
        _direction = direction;
        NSMutableArray<NSNumber *>*participantsList = @[].mutableCopy;
        NSMutableDictionary<NSString *, CallParticipant*>*participantsDictionary = @{}.mutableCopy;
        NSNumberFormatter *formatter = [NSNumberFormatter new];
        formatter.numberStyle = NSNumberFormatterNoStyle;
        
        for (NSNumber *memberId in members.allKeys) {
            NSString *name = members[memberId];
            CallParticipant *participant = [[CallParticipant alloc] initWithParticipantId:memberId fullName:name];
            participantsDictionary[memberId] = participant;
            [participantsList addObject:memberId];
        }
        Profile *profile = [[Profile alloc] init];
        NSNumber *localId = @(profile.ID);
        CallParticipant *local =
        [[CallParticipant alloc] initWithParticipantId:localId fullName:profile.fullName];
        participantsDictionary[localId] = local;
        [participantsList addObject:localId];
        
        _cache = participantsDictionary.copy;
        _list = participantsList.copy;
        _profile = profile;
    }
    return self;
}

// MARK: - Public Methods
- (CallParticipant *)participantWithId:(NSNumber *)userId {
    return self.cache[userId];
}

- (void)updateWithMembers:(NSDictionary<NSNumber *,NSString *> *)members {
    for (NSNumber *userId in members.allKeys) {
        CallParticipant *participant = self.cache[userId];
        if (![participant.fullName isEqualToString:members[userId]] && self.onUpdatedParticipant) {
            self.onUpdatedParticipant(userId, members[userId]);
        }
        participant.fullName = members[userId];
    }
}

// MARK: - QBRTCClientDelegate
- (void)session:(__kindof QBRTCBaseSession *)session
didChangeConnectionState:(QBRTCConnectionState)state
        forUser:(NSNumber *)userID {
    if (![self isCurrentSession:session] || !self.cache[userID]) {
        return;
    }
    CallParticipant *participant = self.cache[userID];
    participant.connectionState = state;
    if (self.onChangedState) {
        self.onChangedState(participant);
    }
}

- (void)session:(__kindof QBRTCBaseSession *)session
updatedStatsReport:(QBRTCStatsReport *)report
      forUserID:(NSNumber *)userID {
    if (![self isCurrentSession:session] || !self.cache[userID]) {
        return;
    }
    if (self.onChangedBitrate) {
        self.onChangedBitrate(userID, report.statsString);
    }
}

//MARK: - Helpers
- (BOOL)isCurrentSession:(__kindof QBRTCBaseSession *)session {
    QBRTCSession *fullSession = (QBRTCSession *)session;
    return [self.callId isEqualToString:fullSession.ID];
}

@end

File: sample-videochat-webrtc-swift/sample-videochat-webrtc-swift/Components/Calls/CallHelper/CallInfo.swift
//
//  CallInfo.swift
//  sample-videochat-webrtc-swift
//
//  Created by Injoit on 24.06.2021.
//  Copyright © 2021 QuickBlox. All rights reserved.
//

import UIKit
import QuickbloxWebRTC

class CallParticipant {
    //MARK: - Properties
    var id: UInt
    var fullName = ""
    var isSelected = false
    var connectionState: QBRTCConnectionState = .new
    var isEnabledSound = true
    
    //MARK: - Life Cycle
    init(id: UInt, fullName: String) {
        self.id = id
        self.fullName = fullName
    }
}

typealias ChangedStateHandler = ( _ participant: CallParticipant) -> Void
typealias ChangedBitrateHandler = ( _ participant: (id: UInt, statsString: String)) -> Void
typealias UpdatedParticipantHandler = ( _ participant: (id: UInt, fullName: String)) -> Void

class CallInfo: NSObject {
    //MARK: - Properties
    var onUpdatedParticipant: UpdatedParticipantHandler?
    var onChangedState: ChangedStateHandler?
    var onChangedBitrate: ChangedBitrateHandler?
    
    private(set) var direction: CallDirection!
    private(set) var callId: String!
    /// The current call participant.
    var localParticipant: CallParticipant? {
        return participant(localParticipantId)
    }
    /// Everyone in the call
    private(set) var participants: [CallParticipant] = []
    /// Someone who participates in a conversation and represents someone else
    private(set) var interlocutors: [CallParticipant] = []
    /// Sorted participants ids.
    private var list: [UInt] = []
    /// The call participants details where the key is a user id.
    private var cache: [UInt : CallParticipant] = [:]
    /// The current user id.
    var localParticipantId: UInt {
        return profile.ID
    }
    private var profile = Profile()

    //MARK: - Life Cycle
    /// The key is a user id and the value is a user name.
    init(callId: String,members: [NSNumber: String], direction: CallDirection) {
        super.init()
        
        self.callId = callId
        self.direction = direction
        var participantsList: [UInt] = []
        var participantsDictionary: [UInt: CallParticipant] = [:]
        var interlocutors: [CallParticipant] = []
        
        for memberId in members.keys {
            let participant = CallParticipant(id: memberId.uintValue, fullName: members[memberId] ?? "Participant")
            participantsDictionary[memberId.uintValue] = participant
            interlocutors.append(participant)
            participantsList.append(memberId.uintValue)
        }
        
        let localParticipantId = profile.ID
        let local = CallParticipant(id: profile.ID, fullName: profile.fullName)
        participantsDictionary[localParticipantId] = local
        participantsList.append(localParticipantId)

        self.interlocutors = interlocutors
        interlocutors.append(local)
        self.participants = interlocutors
        cache = participantsDictionary
        list = participantsList
        
        QBRTCClient.instance().add(self)
    }
    
    //MARK: - Public Methods
    func updateWithMembers(_ members: [NSNumber: String]) {
        for userId in members.keys {
            guard let participant = cache[userId.uintValue],
                  let fullName = members[userId] else {continue}
            participant.fullName = fullName
            onUpdatedParticipant?((userId.uintValue, fullName))
        }
    }
    
    func participant(_ userID: UInt) -> CallParticipant? {
        guard let participant = cache[userID] else {return nil}
        return participant
    }
}

//MARK: - QBRTCClientDelegate
extension CallInfo: QBRTCClientDelegate {
    func session(_ session: QBRTCBaseSession, didChange state: QBRTCConnectionState, forUser userID: NSNumber) {
        if isCurrentSession(session) == false, userID.uintValue != localParticipantId  {
            return
        }
        guard let participant = cache[userID.uintValue] else {
            return
        }
        if participant.connectionState == state {
            return
        }
        participant.connectionState = state
        onChangedState?(participant)
    }
    
    func session(_ session: QBRTCBaseSession, updatedStatsReport report: QBRTCStatsReport, forUserID userID: NSNumber) {
        guard cache[userID.uintValue] != nil else {
            return
        }
        onChangedBitrate?((userID.uintValue, report.statsString()))
    }
    
    func isCurrentSession(_ session: QBRTCBaseSession) -> Bool {
        guard let qbSession = session as? QBRTCSession,
              qbSession.id == callId else { return false }
        return true
    }
}
